### ISSUE TITLE
feat(versions): support patch: versions

### ISSUE THREAD

**Issue by Aghassi** [JamieMason/syncpack#109]
## Description

Currently if you have different packages in a monorepo, and they aren't shared by a workspace, you need to copy paste any patches you create between them. Having a way to enforce that these patches exists and are all the same for any packages that need them would be a helpful check to keeping everything in sync.

## Suggested Solution

Add a `patches` boolean that is defaulted to false, but becomes true in the next major, that let's users opt into this feature. It should respect the grouping that we already do also in the config (if there are multiple versions).

**Comment by JamieMason** [comment]
Can you expand on this please @Aghassi? I don't understand so far, what's a patch in this context?

**Comment by Aghassi** [comment]
@JamieMason With this package https://www.npmjs.com/package/patch-package you can locally patch NPM packages if you are waiting on upstream changes, or need to modify the logic to support your use case that an upstream maintainer doesn't want to take in. This package adds `.patch` files (basically a `git diff` output to a file) into a `patches` directory in a standard format (including version information). So in a repo where there are multiple packages that use the same patches, I thought it might be useful to have alignment with them. Alternatively, I can just work with the flags `patch-package` provides to consolidate patches to a single directory and make sure all users point to that directory.

**Comment by JamieMason** [comment]
It looks like this is something officially supported in yarn and pnpm and there's a Yarn example at https://github.com/ryanhs/yarn2-patch-example/blob/master/package.json. 

Because they're defined using version numbers, it's in syncpack's remit I suppose – I must admit though I'm not motivated to work on this so I'll leave it in as a low priority to look at after some of the other features have landed. Not many users use patches, but this is quite a cool thing.

**Comment by Aghassi** [comment]
Yeah I think this is definitely a case of OSS vs Enterprise. We use patches heavily because we may be waiting on an upstream PR and need to move forward while maintaining the fix. Patches allow us to check in the fix without forking the upstream repo. I can hep contribute to this if you can point me to the relevant code spots. But I agree that it's not _major_, just an edge case I ran into

**Comment by JamieMason** [comment]
Do you have any thoughts how you would resolve mismatches in these @Aghassi? In the above example I see `superagent": "patch:superagent@6.1.0#patches/superagent.patch"` – is that format standardised do you know? or could it vary a lot between projects? I have an idea for how to tackle #112 which I think could sort this as well as a bonus, but if the format of patches is non-deterministic then maybe not.

Thanks.

**Comment by Aghassi** [comment]
I haven't yet tried with pnpm but the API surface between `yarn patch` and `pnpm patch` look very similar if not identical (https://pnpm.io/cli/patch-commit & https://pnpm.io/cli/patch-commit).

I believe there are two scenarios:
1. You don't have a `workspace` and therefore you can't have two versions resolved. If you find two keys for the same package, probably best to error out and ask the user to flatten the dep graph.
2. You don't have `workspaces`, but you are in a repo with multiple `package.json`. You could have different versions here and if you want alignment that's up to the repo owner. This case is hard and I don't have an opinion on how best to handle it.
3. You have `workspaces` which means you can have nested `package.json` each of which _could_ have a different patch version. In this case, we probably want to still align on the version of the dependency as the patch file doesn't have the version information in it (or shouldn't?). I'd say it is also best to add a config option to disable this feature or gate it at first so that you can weed out edge cases. 

I am biased towards `3.` because that's my scenario and I don't want multiple versions of anything unless I am using the `versionGroups` feature, in which case I would guess that the patch linting behaves the same. In general, I want one version.

I'm not sure (yet) if the syntax is standard. I'm migrating to pnpm now internally so I will have a better sense as I migrate. I agree lintable but not fixable is a good start, as well as bias towards erroring not warning.

Thoughts?

**Comment by JamieMason** [comment]
Released in [syncpack@10.0.0](https://github.com/JamieMason/syncpack/releases/tag/10.0.0), please let me know if you run into any issues. `patch:` packages will no longer be ignored, if they are identical they will be valid, if they mismatch they will be reported but they are not auto-fixable.

Note that for nested workspaces there is also https://github.com/JamieMason/syncpack/issues/133 which is not yet complete.
